ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/SourceBuilder.java

```diff
@@ -318,6 +345,60 @@ private static PhysicalSchema getPhysicalSchema(final SourceStep<?> streamSource
         .transformValues(new AddKeyAndTimestampColumns<>(keyGenerator));
   }
 
+  private static void registerChangelogSourceTopic(
+      final String name,
+      final KsqlQueryBuilder queryBuilder,
+      final String sourceSubject
+  ) {
+    final String changelogTopic = changelogTopic(queryBuilder, name);
+    final SchemaRegistryClient srClient = queryBuilder
+        .getServiceContext()
+        .getSchemaRegistryClient();
+    final KafkaTopicClient topicClient = queryBuilder.getServiceContext().getTopicClient();
+
+    try {
+      final Collection<String> allSubjects = srClient.getAllSubjects();
+      // if the source subject isn't using schema registry, don't register a
+      // schema; also if this topic actually exists, we shouldn't add the "fake"
+      // schema to it
+      if (!allSubjects.contains(sourceSubject) || topicClient.isTopicExists(changelogTopic)) {
+        return;
+      }
+
+      final List<Integer> allVersions = srClient.getAllVersions(sourceSubject);
+      for (final Integer version : allVersions) {
+        final Schema schema = srClient.getByVersion(sourceSubject, version, false);
+        final ParsedSchema parsedSchema = srClient.getSchemaById(schema.getId());
+        srClient.register(changelogTopic + SCHEMA_REGISTRY_VALUE_SUFFIX, parsedSchema);
+      }
+    } catch (IOException | RestClientException e) {
+      throw new KsqlException(
```

Comment:
What happens if there are two versions of the source subject (say, 1 and 2), and ksqlDB registers 1 under the changelog subject but hits a transient error while registering 2? Then when ksqlDB tries to start up again, ksqlDB will never try to re-register 2 since the changelog subject exists in Schema Registry. Does this mean we'd still hit the bug for records that use version 2?

Reply:
> Then when ksqlDB tries to start up again, ksqlDB will never try to re-register 2 since the changelog subject exists in Schema Registry

That is not the case, the check is not whether or not the changelog subject exists, but rather whether the _topic_ exists. Every time we will register the schemas under the changelog subjects (which will be a no-op if the schemas are already registered)

Reply:
Ah, brain slip -- sorry about that. What happens if the user registers a new source subject and starts producing records using the new subject version. Will ksqlDB fail to process those records until the next time ksqlDB is restarted (and the new subject version is registered under the changelog subject)?

Reply:
ksqlDB will be able to handle those records fine (just as they are today) because it will use ksql's SerDe to serialize/deserialize from the state store. the problem this PR prevents only happens on recovery, and specifically on recovery when the state store is not on disk - in that case, the bytes from the source topic don't go through the serde and just get piped directly into the state store